components/ai_chat/core/common/mojom/ai_chat.mojom

```diff
@@ -148,6 +148,22 @@ struct ModelWithSubtitle {
   string subtitle;
 };
 
+union ActionDetails {
+  ActionType type;
+  // An empty field indicates that this is not a valid action and will be used for other purposes, such as a subheading.
+  bool empty;
```

Comment:
This seems weird - so this `ActionDetails` type is either a boolean or an Action? I've never seend (or imagined) a `union` used like this. I'm wondering whether you could either have `action_details` be optional? That seems to be the purpose of this union?

When we discussed unions, it was under the assumption that hierarchical categories is what you wanted, instead of a flattened list of items and category headings.

Something more like:

```
union Action {
  ActionsList category;
  LabeledAction action;
};

struct LabeledAction {
  string label;
  Action action;
};

struct ActionsList {
  string category;
  array<Action> actions;
};
```

In that you have an array of items (`ActionList`) in which each item is either a direct `LabeledAction` or a category (`ActionList` again) with it's own children of `LabeledAction` or `ActionList` items, etc. etc.

Also seems strange that the categories have entries in the `ActionType` enum, when they are not selectable.

Reply:
my initial attempt was to have a void/monostate type field but eventually thought it was wasteful on space and ended up with a boolean field to denote if it's "empty"

```
struct Nothing {};
enum Something { A, B, C };

union Foo {
  Nothing nothing;
  Something something;
};
```

the whole goal to have an empty variant was to avoid using optional enums as objective-c runtime (iOS) does not have concept of nullable enum.

Reply:
Sorry, not sure if I'm following this - so in Objective C land we can't have a parameter which is an optional enum?

`SubmitHumanConversationEntry` seems to be taking an optional `ActionType` at the moment - does that need to be changed too?

Reply:
yes, it would have to be marked `-1` or `nan` i guess. i will cross that bridge when fixing iOS CI and will tag the iOS team. for now, i will refactor this mess.

Reply:
i think our objc mojo bindings generator need to be updated to support optional numerics. upstream did this for various bindings generators [here](https://issues.chromium.org/issues/40489779) so we can use that as a guide.

Reply:
I haven't read through the discussion, just a quick comment regarding optional enum, you could define a struct to wrap the primitive value you want inside and make that struct optional.

Reply:
ok, so the void/monotype state, which isn't readable, and the whole n-level nesting of the menu list, make the frontend more complicated with recursive filtering and rendering. also, there will never be that much nesting.

so, i have changed this to use an optional struct which a wrapper to the ActionType enum and this should on iOS too.